docker/runner.go

```diff
@@ -22,52 +29,99 @@ func NewRunner() *Runner {
 }
 
 // Run the function at path
-func (n *Runner) Run(f bosonFunc.Function) error {
-	// Check for the docker binary explicitly so that we can return
-	// an extra-friendly error message.
-	_, err := exec.LookPath("docker")
+func (n *Runner) Run(ctx context.Context, f bosonFunc.Function) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
 	if err != nil {
-		return errors.New("please install 'docker'")
+		return errors.Wrap(err, "failed to create docker api client")
 	}
 
-	if f.Image == "" {
-		return errors.New("Function has no associated Image. Has it been built?")
+	envs := make([]string, 0, len(f.EnvVars)+1)
+	for name, value := range f.EnvVars {
+		if !strings.HasSuffix(name, "-") {
+			envs = append(envs, fmt.Sprintf("%s=%s", name, value))
+		}
+	}
+	if n.Verbose {
+		envs = append(envs, "VERBOSE=true")
 	}
 
-	// Extra arguments to docker
-	args := []string{"run", "--rm", "-t", "-p=8080:8080"}
+	ports := map[nat.Port][]nat.PortBinding{
+		nat.Port("8080/tcp"): {
+			nat.PortBinding{
+				HostPort: "8080",
+				HostIP:   "127.0.0.1",
+			},
+		},
+	}
 
-	for name, value := range f.EnvVars {
-		if !strings.HasSuffix(name, "-") {
-			args = append(args, fmt.Sprintf("-e%s=%s", name, value))
+	conf := &container.Config{
+		Env:          envs,
+		Tty:          false,
+		AttachStderr: true,
+		AttachStdout: true,
+		AttachStdin:  false,
```

Comment:
Only output from function is forwared, no input is sent to the function. I think this is OK. User is not expected to receive anything on stdin anyway. Output is useful for logging.